### ISSUE TITLE
How can I optimize my training speed ?

### ISSUE THREAD

**Issue by bansungmin** [Microsoft/CNTK#1793]
I have about 14 GB feature set (8264220 frames x 440 dimension x 4 byes). 
To speed up training, I am finding optimal parameter settings. 
Training time is compared according to data reading method configuration (minibatchSize=4096) as follows:

1) readMethod=blockRandomize, randomize=1600000, 1GPU
     7 min 41 sec
2) readMethod=blockRandomize, randomize=auto, 1GPU
     3 min 42 sec
3) readMethod=rollingWindow, 1GPU
     4 min 31 sec
4) readMethod=rollingWindow, 4GPU
    17 min 59 sec
5) readMethod=blockRandomize, randomize=auto, 4GPU
     2 min 16 sec

I have enough memory space (132 GB) and 4 GPU(11GB each GPU), which training configuration is best?
In case of using 1 GPU in training, the 2nd configuration (readMethod=blockRandomize, randomize=auto) is the best. 
In the 5th configuration, speed up is not enough even though multiple GPUs are used. 
In that case, GPU usage is fluctuated in the range of 0% to 60%.
Before each epoch begins, it is paging in all features and GPU usage is 0%. 
Even in the each epoch, paging in/out is very frequently occured, then GPU usage is falled down to 0%. 
In the 2nd configuration paging is performed only once before training.
After then, paging in/out is never occured in the each epoch, so GPU usage is maintained to 50~60 %. 
Only difference between 2nd and 5th configuration is whether muliple  or singlue GPU training.
This issue is bottlneck to speed up training.
Somebody has any idea what's wrong?

Thank you.

**Comment by avader906** [comment]
@bansungmin Clearly - given the GPU usage - the bottleneck is getting data into GPUs (https://github.com/microsoft/cntk/wiki/BrainScript-and-Python---Understanding-and-Extending-Readers)
Doesnt it look like most of the time is spent indexing samples rather than in model training (#1435)? 
Can ```keepDataInMemory``` be set to ```true``` - what is the size of the dataset? 
What happens if you set ```randomize = false``` ?

**Comment by bansungmin** [comment]
It seems that setting keepDataInMemory to true is not working, 
still paging in is occured before every epoch begins. This spent most of time.
In the middle of epoch, paging in/out is occured many times and GPU usage is 0%. 
Training with 1 GPU spent 3 min 47 sec, and 8 GPU (2 machines x 4 GPU each) spent 1min 24sec.
(only 2.7 times faster).
The size of dataset is 14 GB (8264220 frames x 440 dimension x 4 byes).
I don't know why paging in all feature set is iterated every epoch inspite of all feature sets are read at first time.

Thank you.

**Comment by avader906** [comment]
@eldakms / team what would be the correct training setup in this use case?

**Comment by eldakms** [comment]
Could you please paste the config + logs here? or send them at eldak at microsoft dot com. I will have a look today. Thank you!

**Comment by bansungmin** [comment]
Config file is pasted below and log file is attached.
Would you please check syncPeriod? 
Various values of syncPeriod are controled, but it doesn't make any changes.
Thank you

### Config

    stderr=$ExpDir$/log
    command=DNN

    precision=float
    parallelTrain=true

    DNN=[
        action=train
        modelPath=$ExpDir$/cntkSpeech.dnn

        # deviceId=-1 for CPU, >=0 for GPU devices
        deviceId=$DeviceNumber$

        traceLevel=1

        # notation xxx:yyy*n:zzz is equivalent to xxx, then yyy repeated n times, then zzz
        # example: 10:20*3:5 is equivalent to 10:20:20:20:5
        NDLNetworkBuilder=[
        ndlMacros=$NdlDir$/default_macros.ndl
            NetworkDescription=$NdlDir$/vad.ndl
        ]

        SGD=[
            epochSize=0
            minibatchSize=4096
            momentumPerMB=0.9
            dropoutRate=0.0
            maxEpochs=10

            learningRatesPerSample=0.002
            parallelTrain = [
                parallelizationMethod = "ModelAveragingSGD"
                distributedMBReading = true

                ModelAveragingSGD = [
                    syncPeriod=5000000
                ]
            ]
        ]

        # Parameter values for the reader
        reader=[
          # reader to use
          readerType=HTKMLFReader

          readMethod=blockRandomize
          randomize=auto

          miniBatchMode=Partial
          verbosity=1

          featIn=[
            dim=440
            scpFile=$WorkDir$/ach/inp.ach.10000
          ]

          featOut_vad=[
            scpFile=$WorkDir$/ach/tgt_vad.ach.10000
            dim=1
          ]
        ]
    ]


### log
[log.txt](https://github.com/Microsoft/CNTK/files/953344/log.txt)

**Comment by bansungmin** [comment]
@avader906 To check feasibility of traing big data using CNTK, small data is trained in advance. 
Connectivity is TCP/IP, disk system throughput is evaluated as below.
In case of training 70 GB data set with 4 GPUs, it seems that 70 x 4 GB memory is used. 
Memory full is occured, and then system is stopped. 
When computing 10 Epochs on 28GB data set, it takes 11 min 13 sec. 
When computing 10 Epochs on 14GB data set, it takes 2 min 16 sec. 

Thank you

### disk system throughput

[user@ ~]$ dd if=/dev/zero iflag=fullblock of=bigfile_1Gx4 bs=4000M count=4
4+0 records in
4+0 records out
16777216000 bytes (17 GB) copied, 14.6689 s, 1.1 GB/s

[user@ ~]$ dd if=/dev/zero iflag=fullblock of=bigfile_1Gx4 bs=400M count=40
40+0 records in
40+0 records out
16777216000 bytes (17 GB) copied, 102.512 s, 164 MB/s

**Comment by eldakms** [comment]
Regarding the sync period - it is a period in samples between two model synchronization. During this period all workers run and update their weights independently, at the end of the period the weights are synchronized. More details regarding this method are given here: 
https://www.microsoft.com/en-us/research/publication/scalable-training-deep-learning-machines-incremental-block-training-intra-block-parallel-optimization-blockwise-model-update-filtering/ 

My understanding is that you want to train in the frame mode, right?
Could you please try out the following configuration for the reader (randomizationWindow=<numberOf Frames> defines how many frames we will prefetch in memory, so you can skip it if you want all data in memory):
...
    reader = [
        verbosity = 0
        randomize = true
        frameMode = true
        randomizationWindow = 8640000

        deserializers = (
            [
                type = "HTKFeatureDeserializer" ; module = "HTKDeserializers"
                input = [
                    features = [
                        dim = 440
                        scpFile = $WorkDir$/ach/inp.ach.10000
                    ]
                ]
            ]:
            [
                type = "HTKFeatureDeserializer" ; module = "HTKDeserializers"
                expandToUtterance = true
                input = [            
                    features = [                        
                        dim = 1
                        scpFile = $WorkDir$/ach/tgt_vad.ach.10000
                    ]
                ]
            ]
        )
    ]

Thank you!

**Comment by bansungmin** [comment]
@eldakms  According to ndl file, "features" are edited. 
When running with new configuration, below error messages are printed.
ach/inp.ach.10000 is as follows:
`O0M30.0.KYD.0335.283.lmf=~/O0M30.0.KYD.0335.283.lmf[0, 665]`

Thank you.

	reader = [
		verbosity = 0
		randomize = true
		frameMode = true
		randomizationWindow = 8640000
		deserializers = (
        [
            type = "HTKFeatureDeserializer" ; module = "HTKDeserializers"
            input = [
                featIn = [
                    dim = 440
                    scpFile = $WorkDir$/ach/inp.ach.10000
                ]
            ]
        ]:
        [
            type = "HTKFeatureDeserializer" ; module = "HTKDeserializers"
            expandToUtterance = true
            input = [            
                featOut_vad = [                        
                    dim = 1
                    scpFile = $WorkDir$/ach/tgt_vad.ach.10000
                ]
            ]
        ]
	    )
	]

### Error message
[CALL STACK]
[0x82139c]
[0x7f8c7a40399f]    Microsoft::MSR::CNTK::HTKDeserializer::  InitializeChunkDescriptions  (std::vector<std::string,std::allocator<std::string>> const&) + 0x9cf
[0x7f8c7a4041e4]    Microsoft::MSR::CNTK::HTKDeserializer::  HTKDeserializer  (std::shared_ptr<Microsoft::MSR::CNTK::CorpusDescriptor>,  Microsoft::MSR::CNTK::ConfigParameters const&,  bool) + 0x7c4
[0x7f8c7a3fd43e]    CreateDeserializer                                 + 0x7e
[0x7f8c7a65284b]    Microsoft::MSR::CNTK::CompositeDataReader::  CreateDeserializer  (Microsoft::MSR::CNTK::ConfigParameters const&,  bool) + 0x1cb
[0x7f8c7a652c7e]    Microsoft::MSR::CNTK::CompositeDataReader::  CreateDeserializers  (Microsoft::MSR::CNTK::ConfigParameters const&) + 0x25e
[0x7f8c7a6535bd]    Microsoft::MSR::CNTK::CompositeDataReader::  CompositeDataReader  (Microsoft::MSR::CNTK::ConfigParameters const&) + 0x74d
[0x7f8c7a65d4e8]                                                       + 0x214e8
[0x7f8ca82f6885]    Microsoft::MSR::CNTK::ReaderShim<float>::  Init  (Microsoft::MSR::CNTK::ConfigParameters const&) + 0x255
[0x7f8ca82c85b7]    Microsoft::MSR::CNTK::DataReader::  DataReader  <Microsoft::MSR::CNTK::ConfigParameters>(Microsoft::MSR::CNTK::ConfigParameters const&) + 0x997
[0x963687]          std::__shared_ptr<Microsoft::MSR::CNTK::DataReader,(__gnu_cxx::_Lock_policy)2>::  __shared_ptr  <std::allocator<Microsoft::MSR::CNTK::DataReader>,Microsoft::MSR::CNTK::ConfigParameters&>(std::_Sp_make_shared_tag,  std
::allocator<Microsoft::MSR::CNTK::DataReader> const&,  Microsoft::MSR::CNTK::ConfigParameters&) + 0x57
[0x9639ae]          std::shared_ptr<Microsoft::MSR::CNTK::DataReader>  CreateObject  <Microsoft::MSR::CNTK::DataReader>(Microsoft::MSR::CNTK::ConfigParameters const&,  wchar_t const*) + 0x2be
[0x96d4d8]          void  DoTrain  <Microsoft::MSR::CNTK::ConfigParameters,float>(Microsoft::MSR::CNTK::ConfigParameters const&) + 0x268
[0x8fe560]          void  DoCommands  <float>(Microsoft::MSR::CNTK::ConfigParameters const&,  std::shared_ptr<Microsoft::MSR::CNTK::MPIWrapper> const&) + 0x9e0
[0x85f7c5]          wmainOldCNTKConfig  (int,  wchar_t**)              + 0x985
[0x85febf]          wmain1  (int,  wchar_t**)                          + 0x7f
[0x810068]          main                                               + 0xd8
[0x7f8ca6254f45]    __libc_start_main                                  + 0xf5
[0x81e64e]
EXCEPTION occurred: Expanded stream should only contain sequences of length 1, utterance 'O0M30.0.KYD.0335.283' has 666

**Comment by eldakms** [comment]
Could you please remove expandToUtterance = true, sorry got a wrong impression that the stream should be expanded? Thanks!

**Comment by bansungmin** [comment]
Error messages have been disappeared. 
However, before each Epoch starts, still GPU usage goes down to 0 %. 
Training with new configuration spent 2 min 18 sec, it is not improved.
Training with previous configuration spent 2 min 16 sec. (2 machines x 4 GPU)
Thank you

**Comment by eldakms** [comment]
Each epoch causes a checkpoint at the end. If the time for a single epoch is 12 seconds probably saving the model to disc for the checkpoint dominates your measurements. 

What is the reported number of samples per second for the runs in the log files? 
Please try to increase the size of your data/epoch, unfortunately we have not yet optimizes for small training data (our usual workloads are 10K-30K hours of speech). In Python you can control checkpointing frequency and can even disable it, in brain script it is not possible.
Thanks!

**Comment by eldakms** [comment]
I have just notices you use ModelAveragingSGD (is there a particular reason to use it?). My comment regarding sync period was regarding BlockMomentum not model averaging, sorry for this.

**Comment by eldakms** [comment]
Please be aware that in general we do not recommend model averaging, more info can be found here:
https://github.com/Microsoft/CNTK/wiki/Multiple-GPUs-and-machines#23-model-averaging-sgd

Thanks!

**Comment by bansungmin** [comment]
You're welcom^^
There isn't particular reaon to use "ModelAveragingSGD". 
I trained it with configuration using "DataParallelSGD" again.
It spent 2 min 44 sec, but GPU usage is so high (about 70~80 %).
Training with "DataParallelSGD" is expected to be improved on big size of data.

Refering to log(above link), training with previous configuration("DataParallelSGD" or "ModelAveragingSGD") spent most time in paging in data before each Epoch starts.
Nevertheless it showed faster speed than that with new configuration.
I think that training with new configuration also spends most time in paging in data.
Because paging in is not printed on log file, I can't check this time consuming.
But after printing below message on log file, GPU usage is 0% in a few second. 

SGD: Saving checkpoint model 'cntkSpeech.dnn.2'
Starting Epoch 3: learning rate per sample = 0.002000  effective momentum = 0.
     900000  momentum as time constant = 38876.0 samples

Log file is attached for samplesPerSecond. 

[log_new_1bit.txt](https://github.com/Microsoft/CNTK/files/955243/log_new_1bit.txt)

**Comment by eldakms** [comment]
When the data is big and randomizationWindow cannot keep everything in memory, loading/unloading chunks at the end of the epoch is expected. But if all can fit in memory then during switch of the epoch there should be no IO except for checkpointing.

Let me double check this and come back to you probably tomorrow. There could be some regression, there were some changes in this area recently.

**Comment by avader906** [comment]
So if the issue is randomization - why not prepare (randomize etc) the data prior to feeding to CNTK?

**Comment by eldakms** [comment]
You have to randomize after each epoch/sweep. Then your corpus  = (initial corpus) * number of epochs. It is not convenient.

**Comment by avader906** [comment]
@eldakms at certain point (your dataset size) you drop convenience in order to achieve time saving for the overall process

**Comment by eldakms** [comment]
Because the randomization is based on chunks and chunk assignment can be different from epoch to epoch between different workers there can be some paging involved at the beginning of the epoch.

If you want to avoid this you can try branch eldak/compositeInMemory with the following config: 
reader = [
    verbosity = 0
    randomize = true
    frameMode = true
    keepDataInMemory = true
    deserializers = (
        [
            type = "HTKFeatureDeserializer" ; module = "HTKDeserializers"
            input = [
                featIn = [
                    dim = 440
                    scpFile = $WorkDir$/ach/inp.ach.10000          # my data here
                ]
            ]
        ]:
        [
            type = "HTKFeatureDeserializer" ; module = "HTKDeserializers"
            input = [            
                featOut_vad = [                        
                    dim = 1
                    scpFile = $WorkDir$/ach/tgt_vad.ach.10000       # my data here
                ]
            ]
        ]
    )
]

As a side note though if you can fit all your data in memory and training takes 12 seconds there is no much point optimizing this toy sample.

Overall, if the GPU is not fully utilized in the middle of the epoch - please scale up the minibatch size. Usually we test with mbSize * number of GPUs you have.
You can enable profiling and see where most of the time is spent, more details how to enable this for BrainScript is here: https://github.com/microsoft/cntk/wiki/BrainScript-and-Python-Performance-Profiler
It could well be that the aggregation takes more than actual processing if the network is simple. Then i would recommend firstly try BlockMomentum and 1BitSGD and if this does not help - build CNTK with NCCL supported (from the logs I see that all 4 GPUs are on the same machine).

Thanks!

**Comment by bansungmin** [comment]
@eldakms In case of using single GPU on single machine, chunk based randomization is performed but paging is never occured after initial paging.
It seemed that training with new configuration still spent most time in paging in data before each Epoch starts. Training efficiency was not improved.
I will enable profiling and analize the result.
Thank you for your help!

**Comment by cha-zhang** [comment]
I'm closing this for now since there doesn't seem any action item remaining from the CNTK team. Please reopen the issue if needed. Thanks!
